float32 dataFloat
string dataString